torch/distributed/_tensor/redistribute.py

```diff
@@ -1,75 +1,132 @@
 # Copyright (c) Meta Platforms, Inc. and affiliates
-from typing import cast, Dict, List, Tuple
+from functools import lru_cache
+from typing import cast, Dict, List, NamedTuple, Tuple
 
 import torch
 import torch.distributed._tensor.api as dtensor
+from torch.distributed._tensor.device_mesh import DeviceMesh
 from torch.distributed._tensor.placement_types import (
     _Partial,
     DTensorSpec,
     Placement,
     Replicate,
     Shard,
 )
-from torch.distributed.device_mesh import DeviceMesh
 
 
-_PlacementItem = Tuple[int, Tuple[Placement, Placement]]
+class _TransformInfo(NamedTuple):
+    mesh_dim: int
+    src_dst_placements: Tuple[Placement, Placement]
+    # logical_shape on this mesh dimension
+    logical_shape: List[int]
 
 
-def _replicate_then_shard(val: _PlacementItem) -> int:
+def _replicate_then_shard(val: _TransformInfo) -> int:
     """
-    Replicate from inner to outer dimension.
-    Shard from outer to inner dimension.
+    Replicate happens from inner to outer dimension.
+    Shard happens from outer to inner dimension.
```

Comment:
Could we add a bit of comment regarding the expected behavior? 

Just want to confirm my understanding is right:
shard src -> partial or replicate => return negative mesh_dim
partial src and replicate src -> dst shard => return postive mesh_dim

And this helps because we need to do replication first, correct?

Reply:
agree. Since I have no background knowledge on `TransformInfo`, this function is not straightforward to people like me.

Reply:
> shard src -> partial or replicate => return negative mesh_dim
partial src and replicate src -> dst shard => return postive mesh_dim

> And this helps because we need to do replication first, correct?

Yes!

I did not change the original approach, but added some more comments for you to understand the details.